Assertion in TIntermUnary::replaceChildNode

lericaa opened the pull request:
#### a74f751c6ec4f129acae8bcfec3a2ecb1c9bccd4
<pre>
Assertion in TIntermUnary::replaceChildNode
<a href="https://rdar.apple.com/123840822">rdar://123840822</a>

Reviewed by NOBODY (OOPS!).

Basic type between kgl_VertexID, kgl_VertexIndex and kgl_VertexIDMetal are different naturally,
Casting kgl_VertexIDMetal(Uint) to kgl_VertexID/kgl_VertexIndex(Int) instead of replace veriable during complie.

* Source/ThirdParty/ANGLE/src/compiler/translator/msl/RewritePipelines.cpp:
(sh::PipelineFunctionEnv::getUpdatedFunction):
(sh::UpdatePipelineFunctions::visitMain):
* Source/ThirdParty/ANGLE/src/compiler/translator/msl/TranslatorMSL.cpp:
(sh::TranslatorMSL::translateImpl):
* Source/ThirdParty/ANGLE/src/compiler/translator/tree_ops/msl/IntroduceVertexIndexID.cpp:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_msl_utils.mm:
(rx::mtl::GenerateTransformFeedbackVaryingOutput):
* Source/ThirdParty/ANGLE/src/tests/compiler_tests/MSLOutput_test.cpp:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a74f751c6ec4f129acae8bcfec3a2ecb1c9bccd4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44202 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23270 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46636 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46846 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40228 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27243 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20664 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36417 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44779 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20337 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38079 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17451 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17819 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39184 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2251 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40412 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39472 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48447 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19181 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15769 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43294 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20542 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42019 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20766 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20168 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/compiler/translator/msl/RewritePipelines.cpp:
```diff
@@ -507,6 +506,14 @@ class PipelineFunctionEnv
                 mPipelineMainLocalVar.external      = instanceIdVar;
                 mPipelineMainLocalVar.externalExtra = baseInstanceVar;
             }
+            else if (isMain && (mPipeline.type == Pipeline::Type::InvocationVertexGlobals))
```
For clarity, could you move this before fragment case? E.g. it would be logical if all the cases handled vertex first, then fragment

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/compiler/translator/msl/RewritePipelines.cpp:
```diff
@@ -808,14 +815,23 @@ class UpdatePipelineFunctions : private TIntermRebuild
                     auto &accessNode = AccessField(*mPipelineMainLocalVar.internal, Name(*var));
-                    auto *assignNode = new TIntermBinary(TOperator::EOpAssign, &accessNode, symbol);
+                    TIntermBinary *assignNode = nullptr;
+                    if (mPipeline.type == Pipeline::Type::InvocationVertexGlobals)
```
This is confusing here as it's in a loop that fragment fields use, too.

Could you move it separately above the fragment case?
also:

In the loop, maybe add an assertion that this is a vertexid qualified variable, so the reader doesn't get surprised that all end up containing the vertex id.

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/compiler/translator/msl/TranslatorMSL.cpp:
```diff
@@ -1138,11 +1132,9 @@ bool TranslatorMSL::translateImpl(TInfoSinkBase &sink,
-                return false;
-            }
-            DeclareRightBeforeMain(*root, kgl_VertexIDMetal);
+            root->insertChildNodes(
```
See void AddFragDepthDeclaration(TIntermBlock &root, TSymbolTable &symbolTable) as an inspiration on how to add this.
So perhaps AddVertexIDDeclaration(..) and then ensure that if client has already  "re"-declared gl_vertexid, don't add another declaration.

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/compiler/translator/msl/TranslatorMSL.cpp:
```diff
@@ -1378,11 +1370,9 @@ bool TranslatorMSL::translateImpl(TInfoSinkBase &sink,
-            DeclareRightBeforeMain(*root, kgl_VertexIDMetal);
+            root->insertChildNodes(
+                FindMainIndex(root),
+                TIntermSequence{new TIntermDeclaration{BuiltInVariable::gl_VertexIndex()}});
```
Maybe here, too, add AddVertexIndexDeclaration

lericaa committed:
Assertion in TIntermUnary::replaceChildNode
rdar://123840822

Reviewed by NOBODY (OOPS!).

Basic type between kgl_VertexID, kgl_VertexIndex and kgl_VertexIDMetal are different naturally,
Casting kgl_VertexIDMetal(Uint) to kgl_VertexID/kgl_VertexIndex(Int) instead of replace veriable during complie.

* Source/ThirdParty/ANGLE/src/compiler/translator/msl/RewritePipelines.cpp:
(sh::PipelineFunctionEnv::getUpdatedFunction):
(sh::UpdatePipelineFunctions::visitMain):
* Source/ThirdParty/ANGLE/src/compiler/translator/msl/TranslatorMSL.cpp:
(sh::TranslatorMSL::translateImpl):
* Source/ThirdParty/ANGLE/src/compiler/translator/tree_ops/msl/IntroduceVertexIndexID.cpp:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_msl_utils.mm:
(rx::mtl::GenerateTransformFeedbackVaryingOutput):
* Source/ThirdParty/ANGLE/src/tests/compiler_tests/MSLOutput_test.cpp:
